### ISSUE TITLE
IllegalArgumentException when using java.util.Date as value for input column

### ISSUE THREAD

**Issue by janhuddel** [camunda/feel-scala#21]
Hi,

I have a simple decision-table with one input-column with type "date". Without this plugin it workes fine. Now, when I activate this plugin I got an IllegalArgumentException from DateDataTypeTransformer. It will try to convert a value with type LocalDateTime to java.util.Date but this fails.

At some point this plugin converts my input from type java.util.Date to LocalDateTime.

Greets
Jan

PS: Version of DmnEngine: 7.7.3

**Comment by saig0** [comment]
Hi @janhuddel , 

where does the input value come from? 
Do you pass the output of another decision into this one as input?
Please share your decision tables / processes.

**Comment by janhuddel** [comment]
Hi,

the input comes directly from process-context (Business-Rule-Task). Is there a way to use this plugin in a standalone DmnEngine without a process? This would be easier for me to create a demo for you.

**Comment by saig0** [comment]
@janhuddel , yes. If you've a standalone engine then you can use the [feel-engine-factory](https://github.com/camunda/feel-scala/tree/master/feel-engine-factory). You can write a test case using the [dmn-engine-test-template](https://github.com/camunda/camunda-engine-dmn-unittest).

Side note: the feel engine uses LocalDateTime instead of Data. So if you have a data-time value that is generated by the feel engine then you've to transform it or don't use the types.

**Comment by saig0** [comment]
Ok, I guess it makes sense to convert date-time values to java.util.Date in a Camunda Context. If so, I can add the transformation to the [CamundaValueMapper](https://github.com/camunda/feel-scala/blob/master/feel-engine-factory/src/main/scala/org/camunda/feel/integration/CamundaValueMapper.scala).

But I'm interested in your use case :)

**Comment by janhuddel** [comment]
Our use case: we have currently a process with a business-rules-task which awaits a date as input. The variable in process context is from type java.util.Date. Without the plugin the transformation works fine.

Now we want to use this plugin to extend the feel-language. After activation of this plugin in out process-engine the rule evaluation fails (without any changes in bpmn- or dmn-file).

This is my testclass: https://gist.github.com/janhuddel/fee376fff160b04cdfe91073d37a845c
Resources: [BPMN](https://gist.github.com/janhuddel/815d712e024d4cb6d0b7631a44303302), [DMN](https://gist.github.com/janhuddel/fcc005e13729548807fcd2de8beb4fe1)

**Comment by saig0** [comment]
Thank you!

The problem is that the feel-engine is used for the input expression and converts the Date object. I'll fix this in the next version.

As a workaround, you can 
* use a String object
* set the expression language of the input expression to 'juel' 
* remove the type 'date'

**Comment by janhuddel** [comment]
Hi,

I've tested my process with the new version (1.3.0). The conversion seems to be correct but now I ran into another problem:
```
Caused by: javax.script.ScriptException: failed to evaluate expression '< date and time("2000-01-01T00:00:00")':
expect Date Time but found 'ValLocalDateTime(2000-01-01T00:00)'
```

I've created a simple process-application so you can easily reprocude that error:
https://github.com/janhuddel/dmn-feel-process.git

Best regards
Jan

**Comment by saig0** [comment]
Hi Jan,

thank you for reaching this out. The situation is not so easy. The problem is that `java.util.Date` is converted to a date-time object in the system time-zone (e.g. with offset +01:00) and it can't be compared to a date-time without offset (the DMN specification says so). 

You can solve the problem by adding an offset. For example:

`< date and time("2000-01-01T00:00:00+01:00")`

Does this work for you?
Is this solution ok for you?

Best regards,
Philipp

**Comment by janhuddel** [comment]
Hm... adding offsets here makes it hard to understand the dmn.
Do you have an answer why this works without the feel-scala-plugin?

**Comment by saig0** [comment]
If you don't want to add an offset to the date-time then you can also transform the date value to a local date-time. For example, using a local date-time variable, a custom data transformer or a proper input expression.  

The Camunda built-in FEEL engine doesn't care about this. Also, it doesn't support all the date-time formats. 

Is this ok for you? 
Do you have a suggestion?

**Comment by janhuddel** [comment]
When I use LocalDateTime instead of Date I got another exception:
```
org.camunda.bpm.dmn.engine.DmnEngineException: DMN-01005 Invalid value '2001-01-17T00:00' for clause with type 'date'.
```
I've added two new tests: testCurrentMillenium_WithLocalDateTime and testCurrentMillenium_WithLocalDate.

The workaround with the string-conversion ("2000-01-01T00:00:00") didn't work anymore with 1.3.0 (in 1.2.0 it is ok) - see testCurrentMillenium_WithString.

I've currently no idea to inject a date to the decision-context.

Or do I understand something completely wrong?

**Comment by saig0** [comment]
Hi Jan,

thank you for the test cases. I try to explain the problem in more detail. 

The first problem is that Camunda doesn't support the new Date-Time API of Java 8 (because it still' based on Java 6). However, if you use Java 8 then you can use the new API. But if you save a variable of the new type then you may notice that these variable types are not supported natively.
Second, the Camunda DMN engine only supports 'java.util.Date' as a date type. That means if you set the input type to 'date' then it tries to convert the value to 'java.util.Date'.
To avoid these problems, I created a [ValueMapper](https://github.com/camunda/feel-scala/blob/master/feel-engine-factory/src/main/scala/org/camunda/feel/integration/CamundaValueMapper.scala) for the FEEL engine plugin. It converts 'LocalDateTime' and 'DateTime' to 'java.util.Date' - every time a value is evaluated. This leads to the problem that you can't use 'LocalDateTime' because it's always converted.

Ok,  how to solve the problems?

* use date-time with an offset `< date and time("2000-01-01T00:00:00+01:00")` (because a 'java.util.Date' is converted to OffsetDateTime)
* use only the date `< date("2000-01-01")` - change the input type to 'local-date' (i.e., something different from 'date') and use 'LocalDate' as variable
* use only the date `< date("2000-01-01")` with input expression `date(varBirthday)` (to extract the date part)- change the input type to 'local-date' (or something else)
* disable the _ValueMapper_ - change the input type to 'date-time' (or something else) and use 'LocalDateTime' as variable 

Is one of these options ok for you?
Do you have questions?

**Comment by janhuddel** [comment]
Hi.

yes, I know that camunda only support java.util.Date. So I prefer the 3rd solution for the provided use-case (check only date part while ignoring the time):
> use only the date < date("2000-01-01") with input expression date(varBirthday) (to extract the date part)- change the input type to 'local-date' (or something else)

But is there any other solution without providing an offset if the time-component is necessarry in the rule?

**Comment by saig0** [comment]
No, currently, it's a bit unhandy. For the next release, I can improve the behavior by
* providing a custom data transformer for the DMN engine to support all FEEL data types (time, local-time, date, date-time, local-date-time, year-month-duration, day-time-duration)
* changing the behavior of the ValueMapper to disable the transformation of (local-)date-time to `java.util.Date` (make it configurable in a way)

Is this ok for you?

**Comment by janhuddel** [comment]
Ok. We'll wait for the next release.
Thanks for your support!

**Comment by saig0** [comment]
Hi Jan,

I've implemented #30 to make it easier to use a date-time expression. I would like to hear your feedback to the solution. 

In your example project, you use an expression `< date and time("2000-01-01T00:00:00")`. You can fix the example by switching to the new version _1.4.0-SNAPSHOT_ and set the input type to `feel-local-date-time` (instead of `date`).

Does it work for you?
Is the solution ok for you?

Best regards,
Philipp
